src/combinatorics/simple_graph/prod.lean

```diff
@@ -165,4 +176,35 @@ by { haveI := (nonempty_prod.1 h.nonempty).1, haveI := (nonempty_prod.1 h.nonemp
 @[simp] lemma box_prod_connected : (G □ H).connected ↔ G.connected ∧ H.connected :=
 ⟨λ h, ⟨h.of_box_prod_left, h.of_box_prod_right⟩, λ h, h.1.box_prod h.2⟩
 
+instance [decidable_eq α] [decidable_eq β] [locally_finite G] [locally_finite H] :
+  locally_finite (G □ H) :=
+begin
+  rintro ⟨_,_⟩,
+  rw box_prod_neighbor_set,
+  apply_instance,
+end
+
+lemma box_prod_degree [decidable_eq α] [decidable_eq β]
+  [locally_finite G] [locally_finite H] (x : α × β) :
+  (G □ H).degree x = G.degree x.1 + H.degree x.2 :=
+begin
+  dsimp only [degree,neighbor_finset],
+  simp_rw [box_prod_neighbor_set x,set.to_finset_union],
+  rw finset.card_disjoint_union,
+  have l : ((λ (a : α), (a, x.2)) '' G.neighbor_set x.1).to_finset.card
+         = (G.neighbor_set x.1).to_finset.card, by
+  { simp only [set.to_finset_card, fintype.card_of_finset,
+               finset.card_image_of_injective _ (prod.mk.inj_right x.2)],},
+  have r : ((λ (b : β), (x.1, b)) '' H.neighbor_set x.2).to_finset.card
+         = (H.neighbor_set x.2).to_finset.card, by
+  { simp only [set.to_finset_card, fintype.card_of_finset,
+               finset.card_image_of_injective _ (prod.mk.inj_left x.1)],},
+  rw [l, r],
+  { rintro ⟨_,_⟩ q,
+    simp only [finset.inf_eq_inter, finset.mem_inter, set.mem_to_finset, set.mem_image,
+               mem_neighbor_set,prod.mk.inj_iff,exists_eq_right_right] at q,
+    obtain ⟨⟨_,_,rfl,rfl⟩,⟨a,_,_,_⟩⟩ := q,
+    exact (a.ne (rfl)).elim, },
+end
```

Comment:
With the suggested `neighbor_set` reformulation, there's a relatively straightforward proof of this lemma.

First, we should have the following lemma added somewhere, likely `data/set/finite.lean`:
```lean
lemma set.to_finset_prod (s : set α) (t : set β) [fintype s] [fintype t] :
  (s ×ˢ t).to_finset = s.to_finset ×ˢ t.to_finset :=
by { ext, simp }
```
Then, the proof of `box_prod_degree` can be simplified to
```lean
lemma box_prod_degree [decidable_eq α] [decidable_eq β]
  [locally_finite G] [locally_finite H] (x : α × β) :
  (G □ H).degree x = G.degree x.1 + H.degree x.2 :=
begin
  simp_rw [← card_neighbor_set_eq_degree, box_prod_neighbor_set,
    ← set.to_finset_card, set.to_finset_union],
  convert finset.card_disjoint_union _;
    simp only [set.to_finset_prod, finset.card_product, set.to_finset_card,
      set.card_singleton, mul_one, one_mul],
  { rintro ⟨_,_⟩ q,
    simp only [finset.inf_eq_inter, finset.mem_inter, finset.mem_product, set.mem_to_finset,
      mem_neighbor_set, set.mem_singleton_iff] at q,
    obtain ⟨⟨q, rfl⟩, ⟨rfl, _⟩⟩ := q,
    exact (q.ne rfl).elim, },
end
```

Reply:
I felt the `finset` lemma would be better inserted just after `theorem to_finset_compl` in `data/fintype.basic.lean`, but I think a finiteness instance is missing at this point… too bad.

Reply:
OK, I'm a bit at a loss for where to put this lemma. For now it's sort of floating in the `finset` section of `data/set/finite.lean`.
What I'd prefer is to move the `fintype` instance for `(s ×ˢ t)` in `data.fintype.basic.lean` and then this lemma there too, but even that is not trivial, because the natural place for the lemma (right after `to_finset_compl`) is actually **before** where the instance would rightfully go.

Reply:
Maybe the best place is right after the `fintype_prod` instance.

Given the definition for `fintype_prod`, I'm wondering whether `rfl` will actually prove this lemma -- could you check?

Reply:
OK, I moved it there.